### ISSUE TITLE
email as username

### ISSUE THREAD

**Issue by mikehedman** [hoodiehq/hoodie-account-server#3]
I should be decided if:
* we presume the username will be an email address
* if not, is an email field required
* if not, we need to make sure there are email arguments for all API's which might need an email address.  I'm looking at:  account.resetPassword(username)  and thinking that processing this request would probably need an email address.

If it's decided that username != email address, then I think we need to add an account.forgotUsername(email) api.

**Comment by gr2m** [comment]
Very good point Mike, I didn't think about this one yet.

I know that other apps provide a feature to receive a username by email, but I wonder if we could let users sign in with both, a username and its email address? Then an `account.forgotUsername` API wouldn't be necessary anymore, right?

**Comment by mikehedman** [comment]
Well, then we loop back to my bullet point 2 in the first message: is email required if id != email.  In order for a user to be able to login with either, and there to always be the failsafe "you forgot your username, so just login with your email this time and go to the Profile page to remind yourself of your username" option, then the user account would need to have a required email field in addition to username.

That being said, yes, it's becoming pretty common for systems to have a "friendly name" username field that is used to describe the user to other users, but then also an email address.  The signin process then checks password and both username and email looking for a match.  In that world, you get for free what you suggest (not needing a forgotUsername() API).  But, that presumes that there's a (validated) email address associated with the account.

**Comment by gr2m** [comment]
Current state is

1. Whether username must be an email address or not is app-specific.
2. Validation rules for user accounts can be defined with the [schema option](https://github.com/hoodiehq/hapi-account/tree/master/server#optionsschema)

   ```js
   // enforce email format for usernames
   schema: {
     username: {
       format: 'email'
     }
   }
   ```

   ```js
   // enforce profile.email property & format   
   schema: {
     profile: {
       type: 'object',
       required: ['email'],
       properties: {
         email: {
           type: 'string',
           format: 'email'
         }
       }
     }
   }

Makes sense?
   ```

**Comment by mikehedman** [comment]
Yeah, this makes sense.  I'll be interested to see the full 'schema' description :)

For the profile one, I might suggest moving the 'required' attribute into the individual properties:
```
schema: {
 profile: {
   type: 'object',
   properties: {
     email: {
       type: 'string',
       format: 'email',
       required: true
     }
   }
 }
}
```

**Comment by gr2m** [comment]
JSON Schema is a thing: http://json-schema.org/ :) I should not assume people know that, I'll link it in the README. 

I'd stick to the JSON Schema, although it brings some overhead with it. I'd only extend it thing things like `unique: true`, that would guarantee that a property value would be unique across all user accounts. That would make sense for `profile.email` for example
